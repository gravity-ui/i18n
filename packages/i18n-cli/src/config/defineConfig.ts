import {ProjectConfig} from './types';

export function defineConfig<Locales extends string[]>(config: ProjectConfig<Locales>) {
    return config;
}
