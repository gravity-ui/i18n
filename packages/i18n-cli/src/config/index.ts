export {defineConfig} from './defineConfig';
export type {ProjectConfig} from './types';
