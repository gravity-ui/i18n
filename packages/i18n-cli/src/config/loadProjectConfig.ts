import {cosmiconfigSync} from 'cosmiconfig';
import {log, MODULE_NAME} from '../shared';
import {ProjectConfig} from './types';

const DEFAULT_LOCALES = ['ru', 'en'];
const DEFAULT_CLIENT_INTL_PATH = 'src/ui/shared/i18n.ts';
const DEFAULT_SERVER_INTL_PATH = 'src/server/utils/i18n.ts';
const DEFAULT_SERVER_PATH_MATCHERS = [/src\/server\/.+$/];

type RequiredConfigOption = 'clientIntlModule' | 'serverIntlModule';

export type NormalizedProjectConfig = Omit<ProjectConfig, RequiredConfigOption> &
    Required<Pick<ProjectConfig, RequiredConfigOption>>;

function normalizeProjectConfig(projectConfig?: ProjectConfig): NormalizedProjectConfig {
    const clientIntlModule = projectConfig?.clientIntlModule;
    const serverIntlModule = projectConfig?.serverIntlModule;

    return {
        ...projectConfig,
        allowedLocales: projectConfig?.allowedLocales || DEFAULT_LOCALES,
        clientIntlModule: {
            ...clientIntlModule,
            path: clientIntlModule?.path || DEFAULT_CLIENT_INTL_PATH,
        },
        serverIntlModule: {
            ...serverIntlModule,
            path: serverIntlModule?.path || DEFAULT_SERVER_INTL_PATH,
            pathMatchers: serverIntlModule?.pathMatchers || DEFAULT_SERVER_PATH_MATCHERS,
        },
    };
}

// Порядок влияет на приоритеты
const DEFAULT_SEARCH_PLACES = ['i18n.config.ts', 'i18n.config.js'];

let cachedProjectConfig: NormalizedProjectConfig | undefined;

export const loadProjectConfig = (searchPlaces?: string[]): NormalizedProjectConfig => {
    if (cachedProjectConfig) {
        return cachedProjectConfig;
    }

    const explorer = cosmiconfigSync(MODULE_NAME, {
        cache: false,
        stopDir: searchPlaces ? undefined : process.cwd(),
        searchPlaces: searchPlaces ?? DEFAULT_SEARCH_PLACES,
    });

    const cfg = explorer.search();

    if (!cfg) {
        log('i18n config not found. Using default values');
    }

    const config = cfg?.config as ProjectConfig | undefined;

    cachedProjectConfig = normalizeProjectConfig(config);
    return cachedProjectConfig;
};
