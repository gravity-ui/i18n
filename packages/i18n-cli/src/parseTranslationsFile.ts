import {
    type Message,
    type MessageMeta,
    type MessageValue,
    type PluralValue,
    type PluralForm,
    isLegacyPlural,
} from '@gravity-ui/i18n-types';
import {MessageWithPlacementMeta, ExportAliases, isTranslationFunction} from './types';
import {MODULE_NAME, removeStartNewLineFromICU} from './shared';
import {simpleTraverse} from '@typescript-eslint/typescript-estree';
import {AST_NODE_TYPES, TSESTree} from '@typescript-eslint/types';
import {parseToAst} from './parsing';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const stringifyObj = (obj: any) => JSON.stringify(obj, null, 4);

export type ParseTranslationsFileArgs = {
    filePath: string;
    content?: string;
};

export type ParseTranslationsFileResult = {
    messages: MessageWithPlacementMeta[];
    filePath: string;
    exportAliases: ExportAliases;
};

function isCreateMessagesCall(node: TSESTree.Node) {
    return (
        node.type === AST_NODE_TYPES.CallExpression &&
        node.callee.type === 'MemberExpression' &&
        node.callee.property.type === 'Identifier' &&
        node.callee.property.name === 'createMessages'
    );
}

function parseStringValue(node: TSESTree.Node): string {
    if (node.type === 'Identifier') {
        return node.name;
    }

    if (node.type === 'Literal' && typeof node.value === 'string') {
        return node.value;
    }

    throw new Error(
        `[${MODULE_NAME}] Node must be string, but get ${node.type} at ${stringifyObj(node.loc)}`,
    );
}

function parseBooleanValue(node: TSESTree.Node): boolean {
    if (node.type === 'Literal' && typeof node.value === 'boolean') {
        return node.value;
    }

    throw new Error(
        `[${MODULE_NAME}] Node must be boolean, but get ${node.type} at ${stringifyObj(node.loc)}`,
    );
}

function parseMessageMeta(node: TSESTree.Node): MessageMeta {
    if (node.type !== 'ObjectExpression') {
        throw new Error(
            `[${MODULE_NAME}] Message meta must be ObjectExpression, but get ${node.type} at ${node.loc}`,
        );
    }

    const meta: MessageMeta = {};

    node.properties.forEach((prop) => {
        if (prop.type !== 'Property') {
            throw new Error(
                `[${MODULE_NAME}] Incorrect node type ${prop.type} at ${stringifyObj(prop.loc)}`,
            );
        }

        const key = parseStringValue(prop.key);

        if (key === 'id') {
            meta.id = parseStringValue(prop.value);
        } else if (key === 'description') {
            meta.description = parseStringValue(prop.value);
        } else if (key === 'markdown') {
            meta.markdown = parseBooleanValue(prop.value);
        } else if (key === 'disableTypograf') {
            meta.disableTypograf = parseBooleanValue(prop.value);
        }
    });

    return meta;
}

function parseMessageValue(node: TSESTree.Node, disallowNesting?: boolean): MessageValue {
    if (node.type === 'Literal') {
        if (node.value === null) {
            return null;
        }

        return parseStringValue(node);
    }

    if (node.type === 'Identifier' && node.name === 'undefined') {
        return null;
    }

    if (node.type === 'TemplateLiteral') {
        if (node.expressions.length || node.quasis.length > 1) {
            throw new Error(
                `[${MODULE_NAME}] Forbidden to use expressions in TemplateLiteral. ${stringifyObj(node.loc)}`,
            );
        }

        return removeStartNewLineFromICU(node.quasis[0]?.value.raw || '');
    }

    if (!disallowNesting && node.type === 'ObjectExpression') {
        const value: PluralValue = {
            one: undefined,
            few: undefined,
            many: undefined,
            zero: undefined,
            two: undefined,
            other: undefined,
        };

        node.properties.forEach((prop) => {
            if (prop.type !== 'Property') {
                throw new Error(
                    `[${MODULE_NAME}] Incorrect node type ${prop.type} at ${stringifyObj(prop.loc)}`,
                );
            }

            const key = parseStringValue(prop.key);

            if (Object.keys(value).includes(key)) {
                const pluralVal = parseMessageValue(prop.value, true);
                if (
                    typeof pluralVal !== 'string' &&
                    typeof pluralVal !== 'undefined' &&
                    pluralVal !== null
                ) {
                    throw new Error();
                }
                value[key as PluralForm] = pluralVal || '';
            } else {
                throw new Error(
                    `[${MODULE_NAME}] Incorrect plural form ${key} at ${stringifyObj(prop.loc)}`,
                );
            }
        });

        return value;
    }

    throw new Error(
        `[${MODULE_NAME}] Incorrect translation value type ${node.type} at ${stringifyObj(node.loc)}`,
    );
}

function parseMessage(node: TSESTree.Node): Message {
    if (node.type !== 'ObjectExpression') {
        throw new Error(
            `[${MODULE_NAME}] Message must be ObjectExpression, but get ${node.type} at ${stringifyObj(node.loc)}`,
        );
    }

    const message: Message = {};

    node.properties.forEach((prop) => {
        if (prop.type !== 'Property') {
            throw new Error(
                `[${MODULE_NAME}] Incorrect node type ${prop.type} at ${stringifyObj(prop.loc)}`,
            );
        }

        const key = parseStringValue(prop.key);

        if (key === 'meta') {
            message.meta = parseMessageMeta(prop.value);
        } else {
            const value = parseMessageValue(prop.value);
            message[key] = value;
        }
    });

    const {pluralCount, normalCount} = Object.entries(message).reduce(
        (acc, [key, cur]) => {
            if (key === 'meta') {
                return acc;
            }

            if (cur === null || cur === '') {
                return acc;
            }

            const isPlural = isLegacyPlural(cur as MessageValue);

            return {
                pluralCount: isPlural ? acc.pluralCount + 1 : acc.pluralCount,
                normalCount: !isPlural ? acc.normalCount + 1 : acc.normalCount,
            };
        },
        {normalCount: 0, pluralCount: 0},
    );

    if (pluralCount > 0 && normalCount > 0) {
        throw new Error(
            `[${MODULE_NAME}] All translations must be strictly in the same form. Problem with message: ${stringifyObj(message)}`,
        );
    }

    return message;
}

function getMessagesFromObjectExpression(
    node: TSESTree.ObjectExpression,
): MessageWithPlacementMeta[] {
    const messages: MessageWithPlacementMeta[] = [];

    node.properties.forEach((prop) => {
        if (prop.type !== 'Property') {
            throw new Error(
                `[${MODULE_NAME}] Incorrect node type ${prop.type} at ${stringifyObj(prop.loc)}`,
            );
        }

        const key = parseStringValue(prop.key);

        messages.push({
            message: parseMessage(prop.value),
            meta: {
                objectKey: key,
            },
        });
    });

    return messages;
}

export async function parseTranslationsFile(
    args: ParseTranslationsFileArgs,
): Promise<ParseTranslationsFileResult> {
    let messages: MessageWithPlacementMeta[] = [];
    const exportAliases: ExportAliases = {};

    const ast = await parseToAst({
        filename: args.filePath,
        content: args.content,
    });

    simpleTraverse(ast, {
        visitors: {
            VariableDeclarator: (node) => {
                if (
                    node.type === AST_NODE_TYPES.VariableDeclarator &&
                    node.id.type === AST_NODE_TYPES.ObjectPattern &&
                    node.init &&
                    node.init.type === AST_NODE_TYPES.CallExpression &&
                    isCreateMessagesCall(node.init)
                ) {
                    node.id.properties.forEach((prop) => {
                        if (
                            prop.type === AST_NODE_TYPES.Property &&
                            prop.key.type === AST_NODE_TYPES.Identifier
                        ) {
                            const originalName = prop.key.name;

                            if (
                                prop.computed === false &&
                                prop.value.type === AST_NODE_TYPES.Identifier &&
                                originalName !== prop.value.name &&
                                isTranslationFunction(originalName)
                            ) {
                                exportAliases[originalName] = prop.value.name;
                            }
                        }
                    });
                }
            },
            CallExpression: (node) => {
                if (
                    node &&
                    node.type === AST_NODE_TYPES.CallExpression &&
                    isCreateMessagesCall(node)
                ) {
                    const arg = node.arguments[0];

                    if (!arg || arg?.type !== 'ObjectExpression') {
                        throw new Error(
                            `[${MODULE_NAME}] Incorrect argument type ${arg?.type} at ${stringifyObj(node.loc)}`,
                        );
                    }

                    messages = getMessagesFromObjectExpression(arg);
                }
            },
        },
    });

    return {
        filePath: args.filePath,
        messages,
        exportAliases,
    };
}
