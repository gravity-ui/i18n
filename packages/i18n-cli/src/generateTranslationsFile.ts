import {
    isLegacyPlural,
    type MessageMeta,
    type MessageValue,
    type PluralValue,
    type Message,
} from '@gravity-ui/i18n-types';
import {writeFile} from 'node:fs/promises';
import {parse, resolve} from 'node:path';
import type {MessageWithPlacementMeta, ExportAliases} from './types';
import {
    createRelativeImport,
    toAbsolutePath,
    TRANSLATIONS_FILE_POSTFIX,
    formatCode,
    isMessageLikeMultilineICU,
    log,
} from './shared';
import {builders as b} from 'estree-toolkit';
import {generate} from 'astring';
import {AssignmentProperty, ObjectExpression, Property} from 'estree';
import parseIcuMessage from 'format-message-parse';
import createPrettyIcuMessage from 'format-message-print';
import {loadProjectConfig, NormalizedProjectConfig} from './config/loadProjectConfig';
import {DEFAULT_EXPORT_ALIASES} from './constants';

export type GenerateTranslationsFileParams = {
    outputPath: string;
    messages: MessageWithPlacementMeta[];
    exportAliases?: ExportAliases;
};

function prettifyMultilineICU(message: string): string {
    try {
        const ast = parseIcuMessage(message);
        return createPrettyIcuMessage(ast);
    } catch (err) {
        log(`Invalid ICU message "${message}"`, err);
        return message;
    }
}

function generateMessageLiteral(message?: string | null) {
    if (message === undefined) {
        return b.identifier('undefined');
    }

    if (message === null) {
        return b.literal(null);
    }

    if (isMessageLikeMultilineICU(message)) {
        const prettyICU = '\n' + prettifyMultilineICU(message);
        return b.templateLiteral(
            [b.templateElement({raw: prettyICU, cooked: prettyICU}, true)],
            [],
        );
    }

    return b.literal(message);
}

function generateLegacyPluralObject(value: PluralValue): ObjectExpression {
    return b.objectExpression(
        Object.entries(value).map(([form, formValue]) =>
            b.property(
                'init',
                b.identifier(form),
                generateMessageLiteral(formValue as string | undefined),
            ),
        ),
    );
}

function generateMeta(meta: MessageMeta): ObjectExpression | undefined {
    const properties: ObjectExpression['properties'] = [];

    if (meta.id) {
        properties.push(b.property('init', b.identifier('id'), b.literal(meta.id)));
    }

    if (meta.description) {
        properties.push(
            b.property('init', b.identifier('description'), b.literal(meta.description)),
        );
    }

    if (typeof meta.markdown === 'boolean') {
        properties.push(b.property('init', b.identifier('markdown'), b.literal(meta.markdown)));
    }

    return properties.length ? b.objectExpression(properties) : undefined;
}

function generateMessageObject(
    message: Message<never>,
    allowedLocales: string[],
): ObjectExpression {
    const propsRecord: Record<string, Property> = {};
    let metaProp: Property | undefined;

    for (const [key, val] of Object.entries(message)) {
        let node: Property['value'] | undefined = undefined;

        if (key === 'meta') {
            const metaObj = generateMeta(val as MessageMeta);
            if (metaObj) {
                metaProp = b.property('init', b.literal('meta'), metaObj);
            }
        } else if (isLegacyPlural(val as MessageValue)) {
            node = generateLegacyPluralObject(val as PluralValue);
        } else if (typeof val === 'string' || val === null) {
            node = generateMessageLiteral(val);
        }

        if (node) {
            propsRecord[key] = b.property('init', b.literal(key), node);
        }
    }

    // Ensure required locales with null if missing
    const sortedProps = allowedLocales.map(
        (locale) => propsRecord[locale] || b.property('init', b.literal(locale), b.literal(null)),
    );

    if (metaProp) {
        sortedProps.push(metaProp);
    }

    return b.objectExpression(sortedProps);
}

function generateMessagesObject(
    messages: MessageWithPlacementMeta[],
    allowedLocales: string[],
): ObjectExpression {
    const sortedMessages = messages
        .slice()
        .sort((a, b) =>
            a.meta.objectKey.localeCompare(b.meta.objectKey, 'en', {sensitivity: 'case'}),
        );

    return b.objectExpression(
        sortedMessages.map(({message, meta}) =>
            b.property(
                'init',
                b.literal(meta.objectKey),
                generateMessageObject(message, allowedLocales),
            ),
        ),
    );
}

function determineIntlModule(outputPath: string, config: NormalizedProjectConfig) {
    return config.serverIntlModule.pathMatchers?.some((m) => m.test(outputPath))
        ? config.serverIntlModule
        : config.clientIntlModule;
}

export function generateTranslationsFileContent(params: GenerateTranslationsFileParams): string {
    const config = loadProjectConfig();

    const intlModule = determineIntlModule(params.outputPath, config);
    const intlImportPath = createRelativeImport(
        intlModule.path,
        params.outputPath,
        intlModule.alias,
    );

    const {exportAliases} = params;
    const tExportName = exportAliases?.t || DEFAULT_EXPORT_ALIASES.t;
    const messageExportName = exportAliases?.Message || DEFAULT_EXPORT_ALIASES.Message;

    const content = generate(
        b.program(
            [
                b.exportNamedDeclaration(
                    b.variableDeclaration('const', [
                        b.variableDeclarator(
                            b.objectPattern(
                                intlModule === config.serverIntlModule
                                    ? [
                                          b.property(
                                              'init',
                                              b.identifier('messages'),
                                              b.identifier('messages'),
                                              false,
                                              true,
                                          ) as AssignmentProperty,
                                      ]
                                    : [
                                          b.property(
                                              'init',
                                              b.identifier(DEFAULT_EXPORT_ALIASES.t),
                                              b.identifier(tExportName),
                                              false,
                                              tExportName === DEFAULT_EXPORT_ALIASES.t,
                                          ) as AssignmentProperty,
                                          b.property(
                                              'init',
                                              b.identifier(DEFAULT_EXPORT_ALIASES.Message),
                                              b.identifier(messageExportName),
                                              false,
                                              messageExportName === DEFAULT_EXPORT_ALIASES.Message,
                                          ) as AssignmentProperty,
                                      ],
                            ),
                            b.callExpression(
                                b.memberExpression(
                                    b.identifier('intl'),
                                    b.identifier('createMessages'),
                                ),
                                [generateMessagesObject(params.messages, config.allowedLocales)],
                            ),
                        ),
                    ]),
                ),
            ],
            'module',
        ),
    );

    return `import {intl} from "${intlImportPath}";\n\n${content}`;
}

function generateOutputFilename(outputPath: string) {
    const absPath = toAbsolutePath(outputPath);
    const parsedPath = parse(absPath);

    if (absPath.endsWith(TRANSLATIONS_FILE_POSTFIX)) {
        return absPath;
    }

    return resolve(parsedPath.ext ? parsedPath.dir : absPath, TRANSLATIONS_FILE_POSTFIX);
}

export async function generateTranslationsFile(params: GenerateTranslationsFileParams) {
    const translationsFilePath = generateOutputFilename(params.outputPath);
    const content = generateTranslationsFileContent(params);

    try {
        const prettyCode = await formatCode(content, translationsFilePath);
        await writeFile(translationsFilePath, prettyCode);
    } catch (err) {
        log(`Failed to generate translations file: ${translationsFilePath}`, err);
        throw err;
    }
}
