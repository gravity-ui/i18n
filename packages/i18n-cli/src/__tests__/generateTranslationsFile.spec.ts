import {generateTranslationsFileContent} from '../generateTranslationsFile';
import {loadProjectConfig} from '../config/loadProjectConfig';
import {parseTranslationsFile} from '../parseTranslationsFile';

jest.mock('../config/loadProjectConfig');

describe('generateTranslationsFileContent', () => {
    it('simple string translations', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en'],
            clientIntlModule: {
                path: 'src/ui/shared/i18n.ts',
                alias: '@shared/i18n',
            },
            serverIntlModule: {
                path: 'src/server/utils/i18n.ts',
            },
        });

        const result = generateTranslationsFileContent({
            outputPath: 'src/ui/units/compute/pages/MainPage',
            messages: [
                {
                    message: {
                        ru: 'О нас',
                        en: 'About us',
                    },
                    meta: {
                        objectKey: 'pageTitle',
                    },
                },
                {
                    message: {
                        ru: 'Имя',
                        en: 'Name',
                    },
                    meta: {objectKey: 'name'},
                },
            ],
        });

        expect(result).toMatchSnapshot();
    });

    it('null in translations', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en'],
            clientIntlModule: {
                path: 'src/ui/shared/i18n.ts',
                alias: '@shared/i18n',
            },
            serverIntlModule: {
                path: 'src/server/utils/i18n.ts',
            },
        });

        const result = generateTranslationsFileContent({
            outputPath: 'src/ui/units/compute/pages/MainPage',
            messages: [
                {
                    message: {
                        'de-DE': null,
                        en: 'About us',
                        'en-IN': null,
                        ru: 'О нас',
                        'ru-KZ': null,
                    },
                    meta: {
                        objectKey: 'pageTitle',
                    },
                },
            ],
        });

        expect(result).toMatchSnapshot();
    });

    it('generate TemplateLiteral for multiline ICU messages', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en'],
            clientIntlModule: {
                path: 'src/ui/shared/i18n.ts',
                alias: '@shared/i18n',
            },
            serverIntlModule: {
                path: 'src/server/utils/i18n.ts',
            },
        });

        const result = generateTranslationsFileContent({
            outputPath: 'src/ui/units/compute/pages/MainPage',
            messages: [
                {
                    message: {
                        en: `this is {count, plural,
                            one{# dog}
                            other{# dogs}
                        }`,
                        ru: `{count, plural,
                            =0 {У нас нет офисов}
                            =1 {У нас всего один офис}
                            other {У нас # офисов}
                        }`,
                    },
                    meta: {objectKey: 'countOffices'},
                },
            ],
        });

        expect(result).toMatchSnapshot();
    });

    it('generate content for server path', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en'],
            clientIntlModule: {
                path: 'src/ui/shared/i18n.ts',
                alias: '@shared/i18n',
            },
            serverIntlModule: {
                path: 'src/server/utils/i18n.ts',
                pathMatchers: [/src\/server\/.+$/], // Default path matchers
            },
        });

        const result = generateTranslationsFileContent({
            outputPath: 'src/server/controllers/payments/main.ts',
            messages: [
                {
                    message: {
                        en: `this is {count, plural,
                            one{# dog}
                            other{# dogs}
                        }`,
                        ru: `{count, plural,
                            =0 {У нас нет офисов}
                            =1 {У нас всего один офис}
                            other {У нас # офисов}
                        }`,
                    },
                    meta: {objectKey: 'countOffices'},
                },
            ],
        });

        expect(result).toMatchSnapshot();
    });

    it('generate content with custom server path matchers', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en'],
            clientIntlModule: {
                path: 'src/ui/shared/i18n.ts',
                alias: '@shared/i18n',
            },
            serverIntlModule: {
                path: 'src/custom-server/shared/intl.ts',
                pathMatchers: [/src\/custom-server\/.+$/],
            },
        });

        const result = generateTranslationsFileContent({
            outputPath: 'src/custom-server/components/tasks/task-controller.ts',
            messages: [
                {
                    message: {
                        en: `this is {count, plural,
                            one{# dog}
                            other{# dogs}
                        }`,
                        ru: `{count, plural,
                            =0 {У нас нет офисов}
                            =1 {У нас всего один офис}
                            other {У нас # офисов}
                        }`,
                    },
                    meta: {objectKey: 'countOffices'},
                },
            ],
        });

        expect(result).toMatchSnapshot();
    });

    it('generate old plural format messages', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en'],
            clientIntlModule: {
                path: 'src/ui/shared/i18n.ts',
                alias: '@shared/i18n',
            },
            serverIntlModule: {
                path: 'src/server/utils/i18n.ts',
            },
        });

        const result = generateTranslationsFileContent({
            outputPath: 'src/ui/units/compute/pages/MainPage',
            messages: [
                {
                    message: {
                        en: {
                            few: undefined,
                            many: undefined,
                            one: 'one',
                            other: 'other',
                            two: undefined,
                            zero: 'zero',
                        },
                    },
                    meta: {objectKey: 'countOffices'},
                },
            ],
        });

        expect(result).toMatchSnapshot();
    });

    it('generate message meta', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en'],
            clientIntlModule: {
                path: 'src/ui/shared/i18n.ts',
                alias: '@shared/i18n',
            },
            serverIntlModule: {
                path: 'src/server/utils/i18n.ts',
            },
        });

        const messages = [
            {
                message: {
                    en: 'About us',
                    ru: 'О нас',
                    meta: {
                        id: 'keyset-test.pageTitle.i984ryg24',
                        description: 'compute page title',
                        markdown: true,
                    },
                },
                meta: {
                    objectKey: 'pageTitle',
                },
            },
        ];

        const generatedContent = generateTranslationsFileContent({
            outputPath: 'src/ui/units/compute/pages/MainPage',
            messages,
        });

        const result = await parseTranslationsFile({
            filePath: 'src/ui/units/compute/pages/MainPage/i18n.ts',
            content: generatedContent,
        });

        expect(result.messages).toEqual(messages);
    });

    it('generate language file in the same directory as the intl', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en'],
            clientIntlModule: {
                path: 'src/ui/shared/i18n.ts',
                alias: '@shared/i18n',
            },
            serverIntlModule: {
                path: 'src/server/utils/i18n.ts',
            },
        });

        const result = generateTranslationsFileContent({
            outputPath: 'src/ui/shared/common.i18n.ts',
            messages: [
                {
                    message: {
                        en: `this is {count, plural,
                            one{# dog}
                            other{# dogs}
                        }`,
                        ru: `{count, plural,
                            =0 {У нас нет офисов}
                            =1 {У нас всего один офис}
                            other {У нас # офисов}
                        }`,
                    },
                    meta: {objectKey: 'countOffices'},
                },
            ],
        });

        expect(result).toMatchSnapshot();
    });

    it('generate meta as last element property in object expression', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en'],
            clientIntlModule: {
                path: 'src/ui/shared/i18n.ts',
                alias: '@shared/i18n',
            },
            serverIntlModule: {
                path: 'src/server/utils/i18n.ts',
            },
        });

        const result = generateTranslationsFileContent({
            outputPath: 'src/ui/shared/common.i18n.ts',
            messages: [
                {
                    message: {
                        meta: {
                            id: 'test-id',
                            description: 'proverka',
                            markdown: true,
                            // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        } as any,
                        en: `this is {count, plural,
                            one{# dog}
                            other{# dogs}
                        }`,
                        ru: `{count, plural,
                            =0 {У нас нет офисов}
                            =1 {У нас всего один офис}
                            other {У нас # офисов}
                        }`,
                    },
                    meta: {objectKey: 'countOffices'},
                },
            ],
        });

        expect(result).toMatchSnapshot();
    });

    it('generate translations for all locales', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en', 'ru-kz'],
            clientIntlModule: {
                path: 'src/ui/shared/i18n.ts',
                alias: '@shared/i18n',
            },
            serverIntlModule: {
                path: 'src/server/utils/i18n.ts',
            },
        });

        const result = generateTranslationsFileContent({
            outputPath: 'src/ui/shared/common.i18n.ts',
            messages: [
                {
                    message: {
                        meta: {
                            id: 'test-id',
                            description: 'proverka',
                            markdown: true,
                            // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        } as any,
                        en: 'proverka',
                    },
                    meta: {objectKey: 'countOffices'},
                },
            ],
        });

        expect(result).toMatchSnapshot();
    });

    it('sort translation keys alphabetically', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en', 'ru-kz'],
            clientIntlModule: {
                path: './i18n.ts',
                alias: './i18n',
            },
            serverIntlModule: {
                path: 'src/server/utils/i18n.ts',
            },
        });

        const result = generateTranslationsFileContent({
            outputPath: 'src/ui/shared/common.i18n.ts',
            messages: [
                {
                    message: {
                        en: 'banana',
                    },
                    meta: {objectKey: 'banana'},
                },
                {
                    message: {
                        en: 'apple',
                    },
                    meta: {objectKey: 'apple'},
                },
                {
                    message: {
                        en: 'cherry',
                    },
                    meta: {objectKey: 'cherry'},
                },
                {
                    message: {
                        en: 'cherry',
                    },
                    meta: {objectKey: 'русский ключ'},
                },
                {
                    message: {
                        en: 'cherry',
                    },
                    meta: {objectKey: 'абвгд'},
                },
            ],
        });

        expect(result).toMatchSnapshot();
    });

    it('generate with aliases in destructuring', async () => {
        jest.mocked(loadProjectConfig).mockReturnValue({
            allowedLocales: ['ru', 'en'],
            clientIntlModule: {
                path: 'src/ui/shared/i18n.ts',
                alias: '@shared/i18n',
            },
            serverIntlModule: {
                path: 'src/server/utils/i18n.ts',
            },
        });

        const result = generateTranslationsFileContent({
            outputPath: 'src/ui/units/compute/pages/MainPage',
            exportAliases: {
                t: 'commonT',
                Message: 'CommonMessage',
            },
            messages: [
                {
                    message: {
                        ru: 'О нас',
                        en: 'About us',
                    },
                    meta: {
                        objectKey: 'pageTitle',
                    },
                },
            ],
        });

        expect(result).toMatchSnapshot();
    });
});
