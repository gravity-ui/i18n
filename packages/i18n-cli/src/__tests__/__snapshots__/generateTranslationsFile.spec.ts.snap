// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateTranslationsFileContent generate TemplateLiteral for multiline ICU messages 1`] = `
"import {intl} from "@shared/i18n";

export const {t, Message} = intl.createMessages({
  "countOffices": {
    "ru": \`
{ count, plural,
     =0 {У нас нет офисов}
     =1 {У нас всего один офис}
  other {У нас # офисов}
}\`,
    "en": \`
this is { count, plural,
    one {# dog}
  other {# dogs}
}\`
  }
});
"
`;

exports[`generateTranslationsFileContent generate content for server path 1`] = `
"import {intl} from "../../utils/i18n";

export const {messages} = intl.createMessages({
  "countOffices": {
    "ru": \`
{ count, plural,
     =0 {У нас нет офисов}
     =1 {У нас всего один офис}
  other {У нас # офисов}
}\`,
    "en": \`
this is { count, plural,
    one {# dog}
  other {# dogs}
}\`
  }
});
"
`;

exports[`generateTranslationsFileContent generate content with custom server path matchers 1`] = `
"import {intl} from "../../shared/intl";

export const {messages} = intl.createMessages({
  "countOffices": {
    "ru": \`
{ count, plural,
     =0 {У нас нет офисов}
     =1 {У нас всего один офис}
  other {У нас # офисов}
}\`,
    "en": \`
this is { count, plural,
    one {# dog}
  other {# dogs}
}\`
  }
});
"
`;

exports[`generateTranslationsFileContent generate language file in the same directory as the intl 1`] = `
"import {intl} from "./i18n";

export const {t, Message} = intl.createMessages({
  "countOffices": {
    "ru": \`
{ count, plural,
     =0 {У нас нет офисов}
     =1 {У нас всего один офис}
  other {У нас # офисов}
}\`,
    "en": \`
this is { count, plural,
    one {# dog}
  other {# dogs}
}\`
  }
});
"
`;

exports[`generateTranslationsFileContent generate meta as last element property in object expression 1`] = `
"import {intl} from "./i18n";

export const {t, Message} = intl.createMessages({
  "countOffices": {
    "ru": \`
{ count, plural,
     =0 {У нас нет офисов}
     =1 {У нас всего один офис}
  other {У нас # офисов}
}\`,
    "en": \`
this is { count, plural,
    one {# dog}
  other {# dogs}
}\`,
    "meta": {
      id: "test-id",
      description: "proverka",
      markdown: true
    }
  }
});
"
`;

exports[`generateTranslationsFileContent generate old plural format messages 1`] = `
"import {intl} from "@shared/i18n";

export const {t, Message} = intl.createMessages({
  "countOffices": {
    "ru": null,
    "en": {
      few: undefined,
      many: undefined,
      one: "one",
      other: "other",
      two: undefined,
      zero: "zero"
    }
  }
});
"
`;

exports[`generateTranslationsFileContent generate translations for all locales 1`] = `
"import {intl} from "./i18n";

export const {t, Message} = intl.createMessages({
  "countOffices": {
    "ru": null,
    "en": "proverka",
    "ru-kz": null,
    "meta": {
      id: "test-id",
      description: "proverka",
      markdown: true
    }
  }
});
"
`;

exports[`generateTranslationsFileContent generate with aliases in destructuring 1`] = `
"import {intl} from "@shared/i18n";

export const {t: commonT, Message: CommonMessage} = intl.createMessages({
  "pageTitle": {
    "ru": "О нас",
    "en": "About us"
  }
});
"
`;

exports[`generateTranslationsFileContent null in translations 1`] = `
"import {intl} from "@shared/i18n";

export const {t, Message} = intl.createMessages({
  "pageTitle": {
    "ru": "О нас",
    "en": "About us"
  }
});
"
`;

exports[`generateTranslationsFileContent simple string translations 1`] = `
"import {intl} from "@shared/i18n";

export const {t, Message} = intl.createMessages({
  "name": {
    "ru": "Имя",
    "en": "Name"
  },
  "pageTitle": {
    "ru": "О нас",
    "en": "About us"
  }
});
"
`;

exports[`generateTranslationsFileContent sort translation keys alphabetically 1`] = `
"import {intl} from "./i18n";

export const {t, Message} = intl.createMessages({
  "apple": {
    "ru": null,
    "en": "apple",
    "ru-kz": null
  },
  "banana": {
    "ru": null,
    "en": "banana",
    "ru-kz": null
  },
  "cherry": {
    "ru": null,
    "en": "cherry",
    "ru-kz": null
  },
  "абвгд": {
    "ru": null,
    "en": "cherry",
    "ru-kz": null
  },
  "русский ключ": {
    "ru": null,
    "en": "cherry",
    "ru-kz": null
  }
});
"
`;
